File: backend/src/modules/auth/auth.dto.ts
import { ApiProperty } from '@nestjs/swagger';
import { IsEmail, IsNotEmpty, IsString, Matches, MinLength } from 'class-validator';

// ==========================================
// 1. 공통 내부 DTO (User, AuthData 등)
// ==========================================

class UserDto {
  @ApiProperty({ example: '550e8400-e29b-41d4-a716-446655440000', description: '사용자 고유 ID' })
  id: string;

  @ApiProperty({ example: 'thief123@example.com', description: '이메일' })
  email: string;

  @ApiProperty({ example: 'lupin_the_third', description: '닉네임' })
  nickname: string;

  @ApiProperty({ example: null, nullable: true, description: '프로필 이미지 URL' })
  profileImage: string | null;
}

class AuthDataDto {
  @ApiProperty({ example: 'eyJhbGciOiJIUzI1Ni...', description: 'JWT Access Token' })
  accessToken: string;

  @ApiProperty({ example: 'dGhwY3ByBcyBhIHJl...', description: 'JWT Refresh Token' })
  refreshToken: string;

  @ApiProperty({ example: 1800, description: 'Access Token 만료 시간 (초)' })
  expiresIn: number;

  @ApiProperty({ type: UserDto, description: '사용자 정보' })
  user: UserDto;
}

class ErrorDetailDto {
  @ApiProperty({ example: 'CONFLICT', description: '에러 코드' })
  code: string;

  @ApiProperty({ example: '이미 존재하는 이메일 또는 닉네임입니다.', description: '에러 메시지' })
  message: string;
}

// ==========================================
// 2. 회원가입 (Signup) 관련
// ==========================================

export class LocalSignupDto {
  @ApiProperty({ example: 'thief123@example.com' })
  @IsEmail()
  email: string;

  @ApiProperty({ example: 'strongPassword123!' })
  @IsString()
  @MinLength(8)
  @Matches(/^(?=.*[a-zA-Z])(?=.*[!@#$%^*+=-])(?=.*[0-9]).{8,25}$/)
  password: string;

  @ApiProperty({ example: 'lupin_the_third' })
  @IsNotEmpty()
  @IsString()
  nickname: string;
}

// ✅ [수정] 명세서와 일치하도록 Response DTO 수정
export class SignupResponseDto {
  @ApiProperty({ example: true, description: '성공 여부' })
  success: boolean;

  @ApiProperty({ example: '회원가입 성공', description: '응답 메시지' })
  message: string;

  @ApiProperty({ type: AuthDataDto, description: '인증 데이터' })
  data: AuthDataDto;

  @ApiProperty({ example: null, nullable: true, description: '에러 정보 (성공 시 null)' })
  error: any;
}

export class SignupConflictErrorDto {
  @ApiProperty({ example: false })
  success: boolean;

  @ApiProperty({ type: ErrorDetailDto })
  error: ErrorDetailDto;
}

// ==========================================
// 3. 로그인 (Login) 관련
// ==========================================

export class LocalLoginDto {
  @ApiProperty({ example: 'thief123@example.com' })
  @IsEmail()
  email: string;

  @ApiProperty({ example: 'strongPassword123!' })
  @IsNotEmpty()
  @IsString()
  password: string;
}

export class KakaoLoginDto {
  @ApiProperty({ example: 'access_token_from_kakao_sdk...', description: '카카오 SDK에서 받은 토큰' })
  @IsNotEmpty()
  @IsString()
  kakaoAccessToken: string;
}

// ✅ [수정] 명세서와 일치하도록 Response DTO 수정
export class LoginResponseDto {
  @ApiProperty({ example: true, description: '성공 여부' })
  success: boolean;

  @ApiProperty({ example: '로그인 성공', description: '응답 메시지' })
  message: string;

  @ApiProperty({ type: AuthDataDto, description: '인증 데이터' })
  data: AuthDataDto;

  @ApiProperty({ example: null, nullable: true, description: '에러 정보 (성공 시 null)' })
  error: any;
}

// ==========================================
// 4. 토큰 재발급 (Refresh)
// ==========================================

export class RefreshRequestDto {
  @ApiProperty({ example: 'dGhwY3ByBcyBhIHJl...', description: 'Refresh Token' })
  @IsNotEmpty()
  @IsString()
  refreshToken: string;
}

// ✅ [수정] 명세서와 일치하도록 Response DTO 수정
export class RefreshResponseDto {
  @ApiProperty({ example: true })
  success: boolean;

  @ApiProperty({ example: '토큰 재발급 성공' })
  message: string;

  @ApiProperty({ 
    example: { 
      accessToken: 'new_access_token...', 
      refreshToken: 'new_refresh_token...' 
    },
    description: '재발급된 토큰 정보'
  })
  data: {
    accessToken: string;
    refreshToken: string;
  };

  @ApiProperty({ example: null, nullable: true })
  error: any;
}

// ==========================================
// 5. 로그아웃 (Logout)
// ==========================================

// ✅ [수정] 명세서와 일치하도록 Response DTO 수정
export class LogoutResponseDto {
  @ApiProperty({ example: true })
  success: boolean;

  @ApiProperty({ example: '로그아웃 성공' })
  message: string;

  @ApiProperty({ example: null, nullable: true })
  error: any;
}

// ==========================================
// 6. 닉네임 중복 확인 (Check Nickname)
// ==========================================

// ✅ [수정] 명세서와 일치하도록 Response DTO 수정
export class CheckNicknameResponseDto {
  @ApiProperty({ example: true })
  success: boolean;

  @ApiProperty({ example: '확인 완료' })
  message: string;

  @ApiProperty({ example: { isAvailable: true }, description: 'true면 사용 가능, false면 중복' })
  data: {
    isAvailable: boolean;
  };

  @ApiProperty({ example: null, nullable: true })
  error: any;
}

// 1. 카카오용 Data DTO (isNewUser 추가됨)
export class KakaoAuthDataDto extends AuthDataDto { // 기존 AuthDataDto 상속
  @ApiProperty({ 
    example: false, 
    description: '신규 가입 유저 여부 (true: 신규, false: 기존)' 
  })
  isNewUser: boolean;
}

// 2. 카카오 로그인 성공 응답 (200 OK)
export class KakaoLoginResponseDto {
  @ApiProperty({ example: true })
  success: boolean;

  @ApiProperty({ example: '카카오 로그인 성공' })
  message: string;

  @ApiProperty({ type: KakaoAuthDataDto }) // 위에서 만든 전용 Data DTO 사용
  data: KakaoAuthDataDto;

  @ApiProperty({ example: null, nullable: true })
  error: any;
}

// 3. 카카오 로그인 실패 응답 (401 Unauthorized)
export class KakaoUnauthorizedErrorDto {
  @ApiProperty({ example: false })
  success: boolean;

  @ApiProperty({ example: '유효하지 않은 카카오 토큰입니다.' })
  message: string;

  @ApiProperty({ example: null, nullable: true })
  data: any;

  @ApiProperty({ 
    example: { code: 'INVALID_KAKAO_TOKEN' },
    description: '에러 상세 정보'
  })
  error: {
    code: string;
  };
}

// ==========================================
// [신규] 토큰 재발급 실패 DTO (401)
// ==========================================
export class RefreshUnauthorizedErrorDto {
  @ApiProperty({ example: false })
  success: boolean;

  @ApiProperty({ 
    type: ErrorDetailDto, 
    description: '에러 상세 정보' 
  })
  @ApiProperty({ 
    example: { 
      code: 'UNAUTHORIZED', 
      message: '유효하지 않거나 만료된 리프레시 토큰입니다. 다시 로그인해주세요.' 
    } 
  })
  error: ErrorDetailDto;
}

File: frontend/lib/data/dto/auth_dto.dart
import 'package:json_annotation/json_annotation.dart';

part 'auth_dto.g.dart';

// ------------------------------------------------------------------
// Shared / Common DTOs
// ------------------------------------------------------------------

@JsonSerializable()
class UserDto {
  final String id;
  final String email;
  final String nickname;
  final String? profileImage;
  final String? provider; // Added to match backend response

  UserDto({
    required this.id,
    required this.email,
    required this.nickname,
    this.profileImage,
    this.provider,
  });

  factory UserDto.fromJson(Map<String, dynamic> json) =>
      _$UserDtoFromJson(json);
  Map<String, dynamic> toJson() => _$UserDtoToJson(this);
}

@JsonSerializable()
class AuthDataDto {
  final String accessToken;
  final String refreshToken;
  final int expiresIn;
  final UserDto user;

  AuthDataDto({
    required this.accessToken,
    required this.refreshToken,
    required this.expiresIn,
    required this.user,
  });

  factory AuthDataDto.fromJson(Map<String, dynamic> json) =>
      _$AuthDataDtoFromJson(json);
  Map<String, dynamic> toJson() => _$AuthDataDtoToJson(this);
}

@JsonSerializable()
class KakaoAuthDataDto extends AuthDataDto {
  final bool isNewUser;

  KakaoAuthDataDto({
    required String accessToken,
    required String refreshToken,
    required int expiresIn,
    required UserDto user,
    required this.isNewUser,
  }) : super(
         accessToken: accessToken,
         refreshToken: refreshToken,
         expiresIn: expiresIn,
         user: user,
       );

  factory KakaoAuthDataDto.fromJson(Map<String, dynamic> json) =>
      _$KakaoAuthDataDtoFromJson(json);
  @override
  Map<String, dynamic> toJson() => _$KakaoAuthDataDtoToJson(this);
}

// ------------------------------------------------------------------
// Signup
// ------------------------------------------------------------------

@JsonSerializable()
class LocalSignupDto {
  final String email;
  final String password;
  final String nickname;

  LocalSignupDto({
    required this.email,
    required this.password,
    required this.nickname,
  });

  Map<String, dynamic> toJson() => _$LocalSignupDtoToJson(this);
}

@JsonSerializable()
class SignupResponseDto {
  final bool? success;
  final String? message;
  final AuthDataDto? data;
  final dynamic error;

  SignupResponseDto({this.success, this.message, this.data, this.error});

  factory SignupResponseDto.fromJson(Map<String, dynamic> json) =>
      _$SignupResponseDtoFromJson(json);
}

@JsonSerializable()
class SignupConflictErrorDto {
  final bool success;
  final dynamic error;

  SignupConflictErrorDto({required this.success, this.error});

  factory SignupConflictErrorDto.fromJson(Map<String, dynamic> json) =>
      _$SignupConflictErrorDtoFromJson(json);
}

// ------------------------------------------------------------------
// Login (Local)
// ------------------------------------------------------------------

@JsonSerializable()
class LocalLoginDto {
  final String email;
  final String password;

  LocalLoginDto({required this.email, required this.password});

  Map<String, dynamic> toJson() => _$LocalLoginDtoToJson(this);
}

@JsonSerializable()
class LoginResponseDto {
  final bool? success;
  final String? message;
  final AuthDataDto? data;
  final dynamic error;

  LoginResponseDto({this.success, this.message, this.data, this.error});

  factory LoginResponseDto.fromJson(Map<String, dynamic> json) =>
      _$LoginResponseDtoFromJson(json);
}

// ------------------------------------------------------------------
// Login (Kakao)
// ------------------------------------------------------------------

@JsonSerializable()
class KakaoLoginDto {
  final String kakaoAccessToken;

  KakaoLoginDto({required this.kakaoAccessToken});

  Map<String, dynamic> toJson() => _$KakaoLoginDtoToJson(this);
}

@JsonSerializable()
class KakaoLoginResponseDto {
  final bool? success;
  final String? message;
  final AuthDataDto? data; // Changed from KakaoAuthDataDto to AuthDataDto
  final bool? isNewUser; // Added root-level field
  final dynamic error;

  KakaoLoginResponseDto({
    this.success,
    this.message,
    this.data,
    this.isNewUser,
    this.error,
  });

  factory KakaoLoginResponseDto.fromJson(Map<String, dynamic> json) =>
      _$KakaoLoginResponseDtoFromJson(json);
}

// ------------------------------------------------------------------
// Refresh
// ------------------------------------------------------------------

@JsonSerializable()
class RefreshRequestDto {
  final String refreshToken;

  RefreshRequestDto({required this.refreshToken});

  Map<String, dynamic> toJson() => _$RefreshRequestDtoToJson(this);
}

@JsonSerializable()
class RefreshResponseData {
  final String accessToken;
  final String refreshToken;

  RefreshResponseData({required this.accessToken, required this.refreshToken});

  factory RefreshResponseData.fromJson(Map<String, dynamic> json) =>
      _$RefreshResponseDataFromJson(json);
}

@JsonSerializable()
class RefreshResponseDto {
  final bool? success;
  final String? message;
  final RefreshResponseData? data;
  final dynamic error;

  RefreshResponseDto({this.success, this.message, this.data, this.error});

  factory RefreshResponseDto.fromJson(Map<String, dynamic> json) =>
      _$RefreshResponseDtoFromJson(json);
}

// ------------------------------------------------------------------
// Logout
// ------------------------------------------------------------------

@JsonSerializable()
class LogoutResponseDto {
  final bool? success;
  final String? message;
  final dynamic error;

  LogoutResponseDto({this.success, this.message, this.error});

  factory LogoutResponseDto.fromJson(Map<String, dynamic> json) =>
      _$LogoutResponseDtoFromJson(json);
}

// ------------------------------------------------------------------
// Check Nickname
// ------------------------------------------------------------------

@JsonSerializable()
class CheckNicknameData {
  final bool isAvailable;

  CheckNicknameData({required this.isAvailable});

  factory CheckNicknameData.fromJson(Map<String, dynamic> json) =>
      _$CheckNicknameDataFromJson(json);
}

@JsonSerializable()
class CheckNicknameResponseDto {
  final bool? success;
  final String? message;
  final CheckNicknameData? data;
  final dynamic error;

  CheckNicknameResponseDto({this.success, this.message, this.data, this.error});

  factory CheckNicknameResponseDto.fromJson(Map<String, dynamic> json) =>
      _$CheckNicknameResponseDtoFromJson(json);
}
